File: backend/src/prechecks/concrete/node/file-descriptor-limit-precheck.ts
import { ElasticClient } from "../../../clients/elastic.client";
import { ConflictError, NotFoundError } from "../../../errors";
import { ClusterNodeType } from "../../../models/cluster-node.model";
import { BaseNodePrecheck } from "../../base/base-node-precheck";
import { ExecutionMode, PrecheckType } from "../../types/enums";
import { NodeContext, PrecheckExecutionRequest } from "../../types/interfaces";
export class FileDescriptorLimitPrecheck extends BaseNodePrecheck {
	private readonly minLimit: number = 65535;

	constructor() {
		super(
			{
				id: "elasticsearch_file_descriptor_check",
				name: "File Descriptor Limit Check",
				type: PrecheckType.NODE,
				mode: ExecutionMode.CODE,
			},
			ClusterNodeType.ELASTIC
		);
	}

	protected async runForContext(request: PrecheckExecutionRequest<NodeContext>): Promise<void> {
		const client = await ElasticClient.buildClient(request.context.node.clusterId);
		const es = client.getClient();

		const nodeId = request.context.node.nodeId;

		const stats = await es.nodes.stats({ node_id: nodeId, metric: "process" });

		const node = stats.nodes[nodeId];
		if (!node) {
			throw new NotFoundError(`Node with ID ${nodeId} not found`);
		}

		const name = node.name;
		const maxFD = node.process?.max_file_descriptors;
		const openFD = node.process?.open_file_descriptors;

		if (!maxFD || !openFD) {
			await this.addLog(request, `${name}: Skipping file descriptor check — missing metrics.`);
			return;
		}

		const usagePercent = (openFD / maxFD) * 100;

		await this.addLog(
			request,
			`${name}: Open FDs = ${openFD}, Max FDs = ${maxFD} (${usagePercent.toFixed(2)}% in use)`
		);

		if (maxFD < this.minLimit) {
			throw new ConflictError(
				`${name}: Max file descriptor limit (${maxFD}) is below the recommended minimum (${this.minLimit}). Consider increasing 'ulimit -n' and systemd LimitNOFILE.`
			);
		}
	}
}

File: server/src/main/java/co/hyperflex/prechecks/concrete/node/elastic/FileDescriptorLimitPrecheck.java
package co.hyperflex.prechecks.concrete.node.elastic;

import co.elastic.clients.elasticsearch.ElasticsearchClient;
import co.elastic.clients.elasticsearch.nodes.NodesStatsResponse;
import co.elastic.clients.elasticsearch.nodes.Process;
import co.elastic.clients.elasticsearch.nodes.Stats;
import co.hyperflex.entities.precheck.PrecheckSeverity;
import co.hyperflex.prechecks.contexts.NodeContext;
import co.hyperflex.prechecks.core.BaseElasticNodePrecheck;
import java.io.IOException;
import java.util.Map;
import org.slf4j.Logger;
import org.springframework.stereotype.Component;

@Component
public class FileDescriptorLimitPrecheck extends BaseElasticNodePrecheck {

  private static final long MIN_LIMIT = 65535;

  @Override
  public String getName() {
    return "File Descriptor Limit Check";
  }

  @Override
  public PrecheckSeverity getSeverity() {
    return PrecheckSeverity.WARNING;
  }

  @Override
  public void run(NodeContext context) {
    String nodeId = context.getNode().getId();
    ElasticsearchClient client = context.getElasticClient().getElasticsearchClient();
    Logger logger = context.getLogger();

    try {
      NodesStatsResponse response = client.nodes().stats(r -> r.nodeId(nodeId).metric("process"));

      Map<String, Stats> nodes = response.nodes();
      Stats node = nodes.get(nodeId);

      if (node == null) {
        throw new RuntimeException("Node with ID [" + nodeId + "] not found in cluster");
      }

      String name = node.name();
      Process process = node.process();

      if (process == null || process.maxFileDescriptors() == null
          || process.openFileDescriptors() == null) {
        logger.info("{}: Skipping file descriptor check — missing metrics.", name);
        return;
      }

      long maxFD = process.maxFileDescriptors();
      long openFD = process.openFileDescriptors();
      double usagePercent = (double) openFD / maxFD * 100;

      logger.info("{}: Open FDs = {}, Max FDs = {} ({}% in use)", name, openFD, maxFD, usagePercent);

      if (maxFD < MIN_LIMIT) {
        logger.error("{}: Max file descriptor limit ({}) is below the recommended minimum ({}).", name, maxFD, MIN_LIMIT);
        logger.error("Consider increasing 'ulimit -n' and systemd LimitNOFILE.");
        throw new RuntimeException();
      }

    } catch (IOException e) {
      logger.error("Failed to retrieve file descriptor stats for node: {}", nodeId, e);
      throw new RuntimeException("Failed to retrieve file descriptor stats for node: " + nodeId, e);
    }
  }
}
